src/spanningtrees/#wilson.jl#

```diff
@@ -0,0 +1,249 @@
+                root[u] = u
+            else
+                nn = outneighbors(G, u)
+                next[u]= rand(nn)
+                u = next[u]
+            end
+        end
+        r = root[u]
+        #Retrace steps, erasing loops
+        u = i
+        while !in_tree[u]
+            root[u] = r
+            in_tree[u] = true
+            u = next[u]
+        end
+    end
+    # If we got here it means we have a single tree, just return it
+    (edges =[Edge{T}(v,next[v]) for v in nodes if next[v] != 0],nroots=1,
+     root=pop!(roots))
+end
+
+
+#maxiter is required by traitfn but pointless for undirected graphs
+@traitfn function random_spanning_tree(G::AG::(!IsDirected); maxiter=-1) where {T,AG <: AbstractGraph{T} }
+    #if (variant==1)
+    ccG = connected_components(G)
+    edges = Array{SimpleEdge{T},1}()
+    roots = zeros(T,length(ccG))
+    for i in eachindex(ccG)
+        r = rand(ccG[i])
+        append!(edges,random_spanning_tree(G,r;nodes=ccG[i],force=true))
+        roots[i] = r
+    end
+    (edges=edges,roots=roots)
+end
+
+
+
+
+@traitfn function random_spanning_tree(G::AG::(IsDirected); maxiter=50) where {T,AG <: AbstractGraph{T} }
+    ccG = connected_components(G)
+    edges = Array{SimpleEdge{T},1}()
+    roots = zeros(T,length(ccG))
+    for i in eachindex(ccG)
+        if (length(ccG[i])>1)
+            tree = _random_spanning_tree_dir(G;nodes=ccG[i],maxiter=maxiter)
+            append!(edges,tree.edges)
+            roots[i] = tree.root
+        else
+            roots[i] = ccG[i][1]
+        end
+    end
+    (edges=edges,roots=roots)
+end
+
+# implementation of Wilson's second algorithm, for directed graphs (outer loop)
+function _random_spanning_tree_dir(G :: AG; nodes=1:nv(G),maxiter=50) where AG <: AbstractGraph{T} where T
+    (sum(outdegree(G)[nodes].==0) > 1) && throw(ArgumentError("This component does not have a spanning tree: there is more than one sink node"))
+    eps = 1.0
+    niter = 1
+    while niter < maxiter #nb: maxiter cut-off needed because you might wait a while
+        eps = eps/2
+        rf=attempt_tree(G,eps;nodes=nodes)
+        if (rf.nroots == 1) #We've found a tree
+            return (edges=rf.edges,root=rf.root)
+        end
+        niter += 1
+    end
+    throw(ErrorException("Failed to find tree, increase maxiter?"))
+end
+
+#given a tree rooted at oldr, root it at newr
+function reorient_tree(G :: AG,oldr :: T,newr :: T) where AG <: AbstractGraph{T} where T
+    (oldr==newr) && return G
+    u = newr
+    iter = 0
+    path = Array{Edge,1}()
+    while u != oldr
+        nn = neighbors(G,u)
+        length(nn) ==0 && throw(ArgumentError("Found unexpected sink node, are you sure you had the right root?")) 
+        next = nn[1]
+        push!(path,Edge(u,next))
+        u = next
+        iter += 1
+        (iter > (ne(G))) && throw(ArgumentError("No path between old root and new, graph is not a spanning tree"))
+    end
+    for e in path
+        rem_edge!(G,e)
```

Comment:
This can be very expensive and in some cases will throw an error since `rem_edge!` isn't defined for all AbstractGraphs.